docs/BinSkimRules.md

```diff
@@ -914,3 +914,25 @@ This rule emits CSV data to the console for every compiler/language/version comb
 
 ---
 
+## Rule `BA6001.DisableIncrementalLinking`
+
+### Description
+
+Incremental linking support increases binary size and can reduce runtime performance. Fully optimized release builds should not specify incremental linking.
```

Comment:
>binary [](http://example.com/codeflow?start=38&length=6)

So, let's maybe add some details on what's going on here, incremental link adds lots of little records, unnecessary data to the PE, yes? Incremental linking disables prevents full optimization (you can't perform whole program LTCG linking for example, right)? I'm guessing that latter issue comprises the greatest perf risk?

Are there other considerations we should mention? an incremental build won't generate a deterministic output, for example. poking around the web, I see a point that the codegen for incremental linking could itself be a source of bugs (idiosyncratic to your non-deterministic build). Makes sense. Anything else?

Reply:
I think the opposite is true.  Enabling /LTCG (or a few other linker options) will prevent incremental linking.  That limits the scope somewhat.  However, I think it is a confusing pitfall that enabling /DEBUG (produce a PDB) also implicitly enables incremental linking.  That is a non-obvious default that in many cases harms performance.  Hence this rule.

I will try to elaborate further in this description.

Reply:
Got it, that definitely argues for a rule. This level of detail is very helpful to include in rule descriptions. This whole nest of command-line args is very involved, thanks for whatever information you can provide.

Reply:
[https://docs.microsoft.com/en-us/cpp/build/reference/incremental-link-incrementally?view=msvc-170#remarks](https://docs.microsoft.com/en-us/cpp/build/reference/incremental-link-incrementally?view=msvc-170#remarks)

"By default, the linker runs in incremental mode. To override a default incremental link, specify /INCREMENTAL:NO.

An incrementally linked program is functionally equivalent to a program that is non-incrementally linked. However, because it is prepared for subsequent incremental links, an incrementally linked executable, static library, or dynamic-link library file:

* Is larger than a non-incrementally linked program because of padding of code and data. Padding enables the linker to increase the size of functions and data without recreating the file.
* May contain jump thunks to handle relocation of functions to new addresses."